modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/ExternalResultHashIndex.java

```diff
@@ -0,0 +1,469 @@
+
+    /** Max HashMap size in entries - capacity. */
+    private long cap;
+
+    /** Total count of the stored entries. */
+    private long entriesCnt;
+
+    /**
+     * @param spillFile File with rows.
+     * @param rowStore External result being indexed by this hash index.
+     * @param initSize Init hash map size.
+     */
+    ExternalResultHashIndex(GridKernalContext ctx, File spillFile, SortedExternalResult rowStore, long initSize) {
+        this.ctx = ctx;
+        dir = spillFile.getParent();
+        spillFileName = spillFile.getName();
+        this.rowStore = rowStore;
+
+        if (initSize <= MIN_CAPACITY)
+            initSize = MIN_CAPACITY;
+
+        long initCap = U.ceilPow2Long(initSize * 2);
+
+        initNewIndexFile(initCap);
+    }
+
+    /**
+     * Puts row into the hash index.
+     *
+     * @param key Row distinct key.
+     * @param rowAddr Row address in the rows file.
+     */
+    public void put(ValueRow key, Long rowAddr) {
+        ensureCapacity();
+
+        int hashCode = key.hashCode();
+
+        putEntryToFreeSlot(hashCode, rowAddr);
+    }
+
+    /**
+     * Finds row address in the hash index by its key.
+     *
+     * @param key Row distinct key.
+     * @return Row address in the rows file.
+     */
+    public Long get(ValueRow key) {
+        Entry entry = findEntry(key);
+
+        return entry == null ? null : entry.rowAddress();
+    }
+
+    /**
+     * Removes row from the hash index.
+     *
+     * @param key Row distinct key
+     * @return Row address in the rows file.
+     */
+    public Long remove(ValueRow key) {
+        Entry entry = findEntry(key);
+
+        if (entry == null)
+            return null;
+
+        writeEntryToIndexFile(entry.slot(), key.hashCode(), -2); // row addr == -2 means removed row.
+
+        entriesCnt--;
+
+        return entry.rowAddress();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void close() throws Exception {
+        U.closeQuiet(fileIo);
+
+        idxFile.delete();
+    }
+
+    /**
+     * Puts new index entry to the empty slot..
+     *
+     * @param hashCode Row key hashcode.
+     * @param rowAddr Row address in the main row store.
+     */
+    private void putEntryToFreeSlot(int hashCode, Long rowAddr) {
+        long slot = findFreeSlotForInsert(hashCode);
+
+        writeEntryToIndexFile(slot, hashCode, rowAddr);
+
+        entriesCnt++;
+    }
+
+    /**
+     * Finds free slot for the given hashcode.
+     *
+     * @param hashCode Hash code.
+     * @return Free slot number.
+     */
+    private long findFreeSlotForInsert(int hashCode) {
+        long slot = slot(hashCode);
+
+        Entry entry;
+
+        while ((entry = readEntryFromIndexFile(slot)) != null && !entry.isRemoved() && !entry.isEmpty()) {
+            slot = slot((int)(slot + 1)); // Check next slot.
+
+            if (slot == slot(hashCode))
+                throw new RuntimeException("No free space left in the hash map.");
```

Comment:
I'm not sure I got you right. Why do you think it is lie? If you didn't find a free slot in the hashmap it means that there is no free space left there. So, it is not a lie.
I may agree that it is better to use `assert` instead of `RuntimeException`.